File: suites/11_ray_tracing/comparisons/photon_mapping.py
def distance_sq(x1: int, y1: int, z1: int, x2: int, y2: int, z2: int) -> int:
    dx, dy, dz = x2 - x1, y2 - y1, z2 - z1
    return dx*dx + dy*dy + dz*dz

def photon_mapping(photons: int) -> int:
    queries = 1024
    checksum = 0
    
    for q in range(queries):
        qx, qy, qz = (q * 17) % 1000, (q * 19) % 1000, (q * 23) % 1000
        gathered = 0
        radius_sq = 10000
        
        for p in range(photons):
            px, py, pz = (p * 31) % 1000, (p * 37) % 1000, (p * 41) % 1000
            d2 = distance_sq(qx, qy, qz, px, py, pz)
            if d2 < radius_sq:
                power = 1000 - (d2 * 1000) // radius_sq
                gathered += power
        checksum = (checksum * 31 + gathered) % 1000000007
    return checksum

if __name__ == "__main__":
    print(photon_mapping(4096))

File: suites/11_ray_tracing/comparisons/photon_mapping.rs
fn distance_sq(x1: i64, y1: i64, z1: i64, x2: i64, y2: i64, z2: i64) -> i64 {
    let dx = x2 - x1; let dy = y2 - y1; let dz = z2 - z1;
    dx*dx + dy*dy + dz*dz
}

fn photon_mapping(photons: i64) -> i64 {
    let queries: i64 = 1024;
    let mut checksum: i64 = 0;
    
    for q in 0..queries {
        let qx = (q * 17) % 1000;
        let qy = (q * 19) % 1000;
        let qz = (q * 23) % 1000;
        
        let mut gathered: i64 = 0;
        let radius_sq: i64 = 10000;
        
        for p in 0..photons {
            let px = (p * 31) % 1000;
            let py = (p * 37) % 1000;
            let pz = (p * 41) % 1000;
            
            let d2 = distance_sq(qx, qy, qz, px, py, pz);
            if d2 < radius_sq {
                let power = 1000 - (d2 * 1000) / radius_sq;
                gathered += power;
            }
        }
        checksum = (checksum * 31 + gathered) % 1000000007;
    }
    checksum
}

fn main() { println!("{}", photon_mapping(4096)); }
